summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/FutureQueue.scala

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2016 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.summingbird.online.executor
+
+import com.twitter.summingbird.online.Queue
+import com.twitter.summingbird.online.option.{ MaxEmitPerExecute, MaxFutureWaitTime, MaxWaitingFutures }
+import com.twitter.util.{ Await, Future, Promise }
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicInteger
+import org.slf4j.{ Logger, LoggerFactory }
+import scala.util.{ Failure, Success, Try }
+
+private[summingbird] object FutureQueue {
+  /**
+   * Ratio of total number of outstanding futures to the portion that is finished
+   * ,at which finished futures are cleared.
+   * Clearing finished futures costs proportional to total number of outstanding
+   * futures, so we want to make sure we only clear when sufficient portion is
+   * finished.
+   */
+  val OutstandingFuturesDequeueRatio = 2
+
+  /**
+   * Wait for n futures to finish. Doesn't block, the returned future is satisfied
+   * once n futures have finished either successfully or unsuccessfully.
+   * If n is greater than number of futures in queue then we wait on all of them.
+   */
+  def waitN[A](fs: Iterable[Future[A]], n: Int): Future[Unit] = {
+    val waitOnCount = Math.min(fs.size, n)
+    if (waitOnCount <= 0) {
+      Future.Unit
+    } else {
+      val count = new AtomicInteger(waitOnCount)
+      val p = Promise[Unit]()
+      fs.foreach { f =>
+        f.ensure {
+          // Note that since we are only decrementing we can cross 0 only
+          // once (unless we decrement more than 2^32 times).
+          if (count.decrementAndGet() == 0) {
+            p.setValue(())
+          }
+        }
+      }
+      p
+    }
+  }
+}
+
+private[summingbird] class FutureQueue[S, T](
+    maxWaitingFutures: MaxWaitingFutures,
+    maxWaitingTime: MaxFutureWaitTime,
+    maxEmitPerExec: MaxEmitPerExecute) {
+  @transient protected lazy val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  private[executor] lazy val outstandingFutures = Queue.linkedNonBlocking[Future[Unit]]
```

Comment:
Tests are currently inspecting this.  It might make sense to modify/eliminate that test (I've broken it for now anyway, I think).  Otherwise it can be private.